api/processors/auxil/algorithms.py

```diff
@@ -0,0 +1,164 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import (
+    GroupProjectStatus,
+    TeacherProjectStatus,
+    TeacherSituationalStatus,
+)
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupBuilderAlgorithm:
+    @dataclass
+    class GroupSizeRestriction:
+        # both min and max student numbers are inclusive:
+        # group size must fit into [min; max]
+        min: int
+        max: int
+
+        def validate_size(self, size: int) -> None:
+            if self.min > size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+            if self.max < size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def __post_init__(self) -> None:
+            # Validate group size against age-related restrictions
+            restrictions = GroupBuilderAlgorithm._get_allowed_group_size(self.age_range)
+            restrictions.validate_size(len(self.students))
+
+    @staticmethod
```

Comment:
@lemontree210 can you please advise if this is correct after merging from main?
We used to have "teaching_accepting_more" status, what's the equivalent going to be? Should I just check maximum number of the groups vs existing number? What statuses should we update for such teacher?

Reply:
> have "teaching_accepting_more" status, what's the equivalent going to be? 
The equivalent will be simply `WORKING`. Those `[NOT]_ACCEPTING_MORE` statuses turned out to be problematic.

> Should I just check maximum number of the groups vs existing number?
Exactly, that's what you should do :)

> What statuses should we update for such teacher?
Their `project_status` does not change until the group starts. The `situational_status` changes to `GROUP_OFFERED` once a tentative group is created, but I think I already did that in #208. So the algorithm just builds a group without changing any statuses, the group creation view should be responsible for that. I don't know if how exactly you're planning to trigger this view, though. Maybe the status-setting functionality from group creation view should be factored out so that both your algorithm and the group creation view can use it.

Reply:
Got it, that's clearer now!

Outstanding question: what do we do with teachers taking on more groups? Situational status still changes to "offered"? Is it ok if they actively work on the first group, but not in "working" status anymore, because they are offered another one?

Reply:
Well that's why I separated `project_status` from `situational_status`: in the situation you're describing (a teacher is working, say, with 1 group but can take on 2 groups in total, so one more group), the teacher who is offered a new group will keep their `project_status=WORKING` but the `situational_status` will change to `GROUP_OFFERED`. Then, once the communication around that second group results in the group either being confirmed or discarded, the `situational_status` is set to `""` again.

Reply:
What if processes happen to 2+ groups at the same time, though? One group wants to set status to AWAITING_START and the other is being GROUP_OFFERED? Can there be any race conditions, groups getting lost, etc?

Reply:
I have thought about it, of course... As far as I understand, right now we're using a synchronous DB engine, so this shouldn't be possible. 

I would also avoid offering two groups to the same teacher at the same time - that could be reflected in an algorithm. I would want the teacher to accept one group and then only after they start teaching, offer another group (if their `simultaneous_groups` is greater than `1`).

This could be done by only adding teacher to a hypothetical group if their `situational_status` is blank (`""`). I guess that even the `TRANSFER_REQUESTED` situational status should preclude a new group from being offered (because a teacher is in a middle of some situation that requires attention).

Reply:
This makes a lot of sense